From dd055178baa466d1aa23206cc6c11c27fda852c8 Mon Sep 17 00:00:00 2001
From: Johannes Berg <johannes.berg@intel.com>
Date: Tue, 17 Mar 2026 17:35:50 +0100
Subject: [PATCH] hostapd: Add require_eht configuration

Add the ability to require EHT, advertising that via the BSS membership
selector as well as rejecting association without EHT.

Signed-off-by: Johannes Berg <johannes.berg@intel.com>
---
 hostapd/config_file.c        |  2 ++
 hostapd/hostapd.conf         |  3 +++
 src/ap/ap_config.h           |  1 +
 src/ap/ieee802_11.c          | 13 +++++++++++++
 src/common/ieee802_11_defs.h |  1 +
 5 files changed, 20 insertions(+)

diff --git a/hostapd/config_file.c b/hostapd/config_file.c
index 97fb044f1..afb983e02 100644
--- a/hostapd/config_file.c
+++ b/hostapd/config_file.c
@@ -4928,6 +4928,10 @@ static int hostapd_config_fill(struct hostapd_config *conf,
 #ifdef CONFIG_IEEE80211BE
 	} else if (os_strcmp(buf, "ieee80211be") == 0) {
 		conf->ieee80211be = atoi(pos);
+		conf->conf_changed = 1;
+	} else if (os_strcmp(buf, "require_eht") == 0) {
+		conf->require_eht = atoi(pos);
+		conf->conf_changed = 1;
 	} else if (os_strcmp(buf, "eht_oper_chwidth") == 0) {
 		conf->eht_oper_chwidth = atoi(pos);
 	} else if (os_strcmp(buf, "eht_oper_centr_freq_seg0_idx") == 0) {
diff --git a/hostapd/hostapd.conf b/hostapd/hostapd.conf
index a09c18bbd..fac07cd37 100644
--- a/hostapd/hostapd.conf
+++ b/hostapd/hostapd.conf
@@ -1057,6 +1057,9 @@ wmm_ac_vo_acm=0
 # 1 = enabled
 #ieee80211be=1
 
+# Require stations to support EHT PHY (reject association if they do not)
+#require_eht=1
+
 #disable_11be: Boolean (0/1) to disable EHT for a specific BSS
 #disable_11be=0
 
diff --git a/src/ap/ap_config.h b/src/ap/ap_config.h
index cb08f4baf..98a2f1bb8 100644
--- a/src/ap/ap_config.h
+++ b/src/ap/ap_config.h
@@ -1230,6 +1230,7 @@ struct hostapd_config {
 	u8 eht_oper_centr_freq_seg0_idx;
 	struct eht_phy_capabilities_info eht_phy_capab;
 	u16 punct_bitmap; /* a bitmap of disabled 20 MHz channels */
+	bool require_eht;
 #endif /* CONFIG_IEEE80211BE */
 
 	/* EHT enable/disable config from CHAN_SWITCH */
diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index 1c01821a2..2f34bd983 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -131,6 +131,10 @@ u8 * hostapd_eid_supp_rates(struct hostapd_data *hapd, u8 *eid)
 	if (hapd->iconf->ieee80211ax && hapd->iconf->require_he)
 		num++;
 #endif /* CONFIG_IEEE80211AX */
+#ifdef CONFIG_IEEE80211BE
+	if (hapd->iconf->ieee80211be && hapd->iconf->require_eht)
+		num++;
+#endif /* CONFIG_IEEE80211BE */
 	h2e_required = (hapd->conf->sae_pwe == 1 ||
 			hostapd_sae_pw_id_in_use(hapd->conf) == 2) &&
 		hapd->conf->sae_pwe != 3 &&
@@ -165,6 +169,13 @@ u8 * hostapd_eid_supp_rates(struct hostapd_data *hapd, u8 *eid)
 	}
 #endif /* CONFIG_IEEE80211AX */
 
+#ifdef CONFIG_IEEE80211BE
+	if (hapd->iconf->ieee80211be && hapd->iconf->require_eht && count < 8) {
+		count++;
+		*pos++ = 0x80 | BSS_MEMBERSHIP_SELECTOR_EHT_PHY;
+	}
+#endif /* CONFIG_IEEE80211BE */
+
 	if (h2e_required && count < 8) {
 		count++;
 		*pos++ = 0x80 | BSS_MEMBERSHIP_SELECTOR_SAE_H2E_ONLY;
@@ -186,6 +197,10 @@ u8 * hostapd_eid_ext_supp_rates(struct hostapd_data *hapd, u8 *eid)
 	if (hapd->iconf->ieee80211ax && hapd->iconf->require_he)
 		num++;
 #endif /* CONFIG_IEEE80211AX */
+#ifdef CONFIG_IEEE80211BE
+	if (hapd->iconf->ieee80211be && hapd->iconf->require_eht)
+		num++;
+#endif /* CONFIG_IEEE80211BE */
 	h2e_required = (hapd->conf->sae_pwe == 1 ||
 			hostapd_sae_pw_id_in_use(hapd->conf) == 2) &&
 		hapd->conf->sae_pwe != 3 &&
@@ -231,6 +246,14 @@ u8 * hostapd_eid_ext_supp_rates(struct hostapd_data *hapd, u8 *eid)
 	}
 #endif /* CONFIG_IEEE80211AX */
 
+#ifdef CONFIG_IEEE80211BE
+	if (hapd->iconf->ieee80211be && hapd->iconf->require_eht) {
+		count++;
+		if (count > 8)
+			*pos++ = 0x80 | BSS_MEMBERSHIP_SELECTOR_EHT_PHY;
+	}
+#endif /* CONFIG_IEEE80211BE */
+
 	if (h2e_required) {
 		count++;
 		if (count > 8)
@@ -4803,6 +4826,14 @@ static int check_assoc_ies(struct hostapd_data *hapd, struct sta_info *sta,
 					  elems.eht_capabilities_len);
 		if (resp != WLAN_STATUS_SUCCESS)
 			return resp;
+
+		if (hapd->iconf->require_eht && !(sta->flags & WLAN_STA_EHT)) {
+			hostapd_logger(hapd, sta->addr,
+				       HOSTAPD_MODULE_IEEE80211,
+				       HOSTAPD_LEVEL_INFO,
+				       "Station does not support mandatory EHT PHY - reject association");
+			return WLAN_STATUS_DENIED_EHT_NOT_SUPPORTED;
+		}
 	}
 #endif /* CONFIG_IEEE80211BE */
 
diff --git a/src/common/ieee802_11_defs.h b/src/common/ieee802_11_defs.h
index 83f9d5c8b..cf8870019 100644
--- a/src/common/ieee802_11_defs.h
+++ b/src/common/ieee802_11_defs.h
@@ -212,6 +212,7 @@
 #ifdef CONFIG_MTK_IEEE80211BE
 #define WLAN_STATUS_INVALID_MLIE 128
 #endif
+#define WLAN_STATUS_DENIED_EHT_NOT_SUPPORTED 135
 
 /* Reason codes (IEEE Std 802.11-2016, 9.4.1.7, Table 9-45) */
 #define WLAN_REASON_UNSPECIFIED 1
@@ -1289,6 +1290,7 @@ struct ieee80211_ampe_ie {
 #define HT_OPER_PARAM_PCO_PHASE				((u16) BIT(11))
 /* B36..B39 - Reserved */
 
+#define BSS_MEMBERSHIP_SELECTOR_EHT_PHY 121
 #define BSS_MEMBERSHIP_SELECTOR_HE_PHY 122
 #define BSS_MEMBERSHIP_SELECTOR_VHT_PHY 126
 #define BSS_MEMBERSHIP_SELECTOR_HT_PHY 127
